engines/sci/graphics/controls32.cpp

```diff
@@ -353,4 +366,303 @@ void GfxControls32::flashCursor(TextEditor &editor) {
+		_bottomVisibleLine = _numLines - 1;
+
+	_firstVisibleChar = _startsOfLines[_topVisibleLine];
+	_lastVisibleChar = _startsOfLines[_bottomVisibleLine+1] - 1;
+
+	_visibleText = Common::String(_text.c_str() + _firstVisibleChar, _text.c_str() + _lastVisibleChar + 1);
+
+	// TODO: Double check this -1 at the end (for the \n)
+	Common::String lineText(_text.c_str() + _startsOfLines[_topVisibleLine], _text.c_str() + _startsOfLines[_topVisibleLine+1] - 1);
+
+	debugC(3, kDebugLevelGraphics, "ScrollWindow::upArrow: top: %d, bottom: %d, num: %d, numvis: %d, lineText: %s", _topVisibleLine, _bottomVisibleLine, _numLines, _numVisibleLines, lineText.c_str());
+
+	_gfxText32->scrollLine(lineText, _numVisibleLines, _foreColor, _alignment, _fontId, kScrollUp);
+
+	if (_visible) {
+		assert(_screenItem);
+
+		_screenItem->update();
+		g_sci->_gfxFrameout->frameOut(true);
+	}
+}
+
+void ScrollWindow::downArrow() {
+	if (_topVisibleLine + 1 >= _numLines)
+		return;
+
+	_topVisibleLine++;
+	_bottomVisibleLine++;
+
+	if (_bottomVisibleLine + 1 >= _numLines)
+		_bottomVisibleLine = _numLines - 1;
+
+	_firstVisibleChar = _startsOfLines[_topVisibleLine];
+	_lastVisibleChar = _startsOfLines[_bottomVisibleLine+1] - 1;
+
+	_visibleText = Common::String(_text.c_str() + _firstVisibleChar, _text.c_str() + _lastVisibleChar + 1);
+
+	Common::String lineText;
+	if (_bottomVisibleLine - _topVisibleLine + 1 == _numVisibleLines) {
+		// TODO: Double check this -1 at the end (for the \n)
+		lineText = Common::String(_text.c_str() + _startsOfLines[_bottomVisibleLine], _text.c_str() + _startsOfLines[_bottomVisibleLine+1] - 1);
+	} else {
+		// scroll in empty string
+	}
+
+	debugC(3, kDebugLevelGraphics, "ScrollWindow::downArrow: top: %d, bottom: %d, num: %d, numvis: %d, lineText: %s", _topVisibleLine, _bottomVisibleLine, _numLines, _numVisibleLines, lineText.c_str());
+
+
+	_gfxText32->scrollLine(lineText, _numVisibleLines, _foreColor, _alignment, _fontId, kScrollDown);
+
+	if (_visible) {
+		assert(_screenItem);
+
+		_screenItem->update();
+		g_sci->_gfxFrameout->frameOut(true);
+	}
+}
+
+
+void ScrollWindow::getLineIndices() {
+	_gfxText32->setFont(_fontId);
+	// set _gfxText32 foreColor, alignment?
+
+	if (_gfxText32->_font->getHeight() != _pointSize) {
+		error("ScrollWindow font size mismatch");
+	}
+
+	Common::Rect r(0, 0, _rect10.width() + 1, _pointSize + 3);
+
+	_startsOfLines.clear();
+
+	int index = 0;
+
+	while (index < (int)_text.size()) {
+		_startsOfLines.push_back(index);
+
+		index += _gfxText32->getTextCount(_text, index, r, false);
+	}
+	_numLines = _startsOfLines.size();
+
+	_startsOfLines.push_back(_text.size());
+
+	_lastVisibleChar = _gfxText32->getTextCount(_text, 0, _fontId, _rect10, false) - 1;
+
+	_bottomVisibleLine = 0;
+	while (_bottomVisibleLine < _numLines - 1 &&
+	       _startsOfLines[_bottomVisibleLine+1] < _lastVisibleChar)
+		_bottomVisibleLine++;
+
+	_numVisibleLines = _bottomVisibleLine + 1;
+}
+
+// Call after changing text or text position
```

Comment:
Can this be put to doxygen comment in the header for this function? “This method must be called after changing the text or text position.”